packages/flutter/lib/src/rendering/auto_layout.dart

```diff
@@ -38,153 +37,154 @@ abstract class _AutoLayoutParamMixin {
-  /// active solver. If no implicit constraints needs to be applied, the object
-  /// may return null.
-  List<al.Constraint> _constructImplicitConstraints();
-
-  void _setupImplicitConstraints(al.Solver solver) {
-    List<al.Constraint> implicit = _constructImplicitConstraints();
+  List<al.Constraint> _implicitConstraints;
 
-    if (implicit == null || implicit.length == 0) {
+  void _addImplicitConstraints() {
+    assert(_renderBox != null);
+    if (_renderBox.parent == null)
       return;
-    }
-
-    al.Result result = solver.addConstraints(implicit);
+    assert(_renderBox.parent is RenderAutoLayout);
+    final RenderAutoLayout parent = _renderBox.parent;
+    final AutoLayoutParentData parentData = _renderBox.parentData;
+    final List<al.Constraint> implicit = parentData._constructImplicitConstraints();
+    if (implicit == null || implicit.isEmpty)
+      return;
+    final al.Result result = parent._solver.addConstraints(implicit);
     assert(result == al.Result.success);
-
+    parent.markNeedsLayout();
     _implicitConstraints = implicit;
   }
 
-  void _removeImplicitConstraints(al.Solver solver) {
-    if (_implicitConstraints == null || _implicitConstraints.length == 0) {
+  void _removeImplicitConstraints() {
+    assert(_renderBox != null);
+    if (_renderBox.parent == null)
       return;
-    }
-
-    al.Result result = solver.removeConstraints(_implicitConstraints);
+    if (_implicitConstraints == null || _implicitConstraints.isEmpty)
+      return;
+    assert(_renderBox.parent is RenderAutoLayout);
+    final RenderAutoLayout parent = _renderBox.parent;
+    final al.Result result = parent._solver.removeConstraints(_implicitConstraints);
     assert(result == al.Result.success);
-
+    parent.markNeedsLayout();
     _implicitConstraints = null;
   }
 }
 
-class AutoLayoutParentData extends ContainerBoxParentDataMixin<RenderBox> with _AutoLayoutParamMixin {
-
-  AutoLayoutParentData(this._renderBox) {
-    _setupLayoutParameters(this);
-  }
+class AutoLayoutParentData extends ContainerBoxParentDataMixin<RenderBox> {
+  AutoLayoutParentData(this._renderBox);
 
   final RenderBox _renderBox;
 
-  void _applyAutolayoutParameterUpdates() {
-    // This is called by the parent's layout function
-    // to lay our box out.
-    assert(_renderBox.parentData == this);
-    assert(() {
-      final RenderAutoLayout parent = _renderBox.parent;
-      assert(parent.debugDoingThisLayout);
-    });
-    BoxConstraints size = new BoxConstraints.tightFor(
-      width: _rightEdge.value - _leftEdge.value,
-      height: _bottomEdge.value - _topEdge.value
+  AutoLayoutParams get params => _params;
+  AutoLayoutParams _params;
+  void set params(AutoLayoutParams value) {
+    if (_params == value)
+      return;
+    if (_params != null) {
+      _params._removeImplicitConstraints();
+      _params._renderBox = null;
+    }
+    _params = value;
+    if (_params != null) {
+      assert(_params._renderBox == null);
```

Comment:
>  Ideally I would be able to associate a context with a parameter when I added to to a solver.

You can already do this on [parameters you add to the solver]( https://github.com/flutter/flutter/blob/cdc4055444dc8f9f9183d0919cbf5e1c74d2c86b/packages/cassowary/lib/param.dart#L9). At that level of interaction with the solver, a `Variable` is the solvers view of an indeterminate in the linear equation while a `Param` maps that indeterminate to some quantity in the outside world. How changes to the `Variable` affect anything in the outside world is not the concern (nor responsibility) of the solver.

The `AutoLayoutParams` object was my (possibly clumsy) attempt at encapsulating all the variables on an entity that could possibly be affected by auto layout. We might want to retool that if you find the current interaction with the solver is not as elegant as it could be.